consensus/src/dag/dag_driver.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    dag::{
+        dag_store::Dag,
+        reliable_broadcast::ReliableBroadcast,
+        types::{CertificateAckState, CertifiedNode, Node, NodeMetadata, SignatureBuilder},
+    },
+    state_replication::PayloadClient,
+    util::time_service::TimeService,
+};
+use aptos_consensus_types::common::{Author, Payload};
+use aptos_infallible::RwLock;
+use aptos_types::{block_info::Round, epoch_state::EpochState};
+use futures::{
+    future::{AbortHandle, Abortable},
+    FutureExt,
+};
+use std::sync::Arc;
+
+struct DagDriver {
+    author: Author,
+    epoch_state: Arc<EpochState>,
+    dag: Arc<RwLock<Dag>>,
+    payload_client: Arc<dyn PayloadClient>,
+    reliable_broadcast: Arc<ReliableBroadcast>,
+    current_round: Round,
+    time_service: Arc<dyn TimeService>,
+    rb_abort_handle: Option<AbortHandle>,
+}
+
+impl DagDriver {
+    pub fn new(
+        author: Author,
+        epoch_state: Arc<EpochState>,
+        dag: Arc<RwLock<Dag>>,
+        payload_client: Arc<dyn PayloadClient>,
+        reliable_broadcast: Arc<ReliableBroadcast>,
+        current_round: Round,
+        time_service: Arc<dyn TimeService>,
+    ) -> Self {
+        Self {
+            author,
+            epoch_state,
+            dag,
+            payload_client,
+            reliable_broadcast,
+            current_round,
+            time_service,
+            rb_abort_handle: None,
+        }
+    }
+
+    pub fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        let mut dag_writer = self.dag.write();
+        let round = node.metadata().round();
+        if dag_writer.all_exists(node.parents().iter().map(|metadata| metadata.digest())) {
+            dag_writer.add_node(node)?;
+            if self.current_round == round {
+                let maybe_strong_links = dag_writer
+                    .get_strong_links_for_round(self.current_round, &self.epoch_state.verifier);
+                drop(dag_writer);
+                if let Some(strong_links) = maybe_strong_links {
+                    self.enter_new_round(strong_links);
+                }
+            }
+        }
```

Comment:
We need an else case here to return an `Err` if `all_exists` is false?

Reply:
yeah we need to handle the fetch case, I'll add a todo